python/distributed-ucxx/distributed_ucxx/ucxx.py

```diff
@@ -90,13 +90,84 @@ def synchronize_stream(stream=0):
     stream.synchronize()
 
 
+def _stop_notifier_thread_and_progress_tasks():
+    """Stop the notifier thread and progress tasks.
+
+    If no Dask resources that make use of UCXX communicator are running anymore,
+    stop the notifier thread and progress tasks to allow for clean shutdown.
+    """
+    ctx = ucxx.core._get_ctx()
+    if len(ctx._dask_resources) == 0:
+        ctx.stop_notifier_thread()
+        ctx.progress_tasks.clear()
+
+
+def _register_dask_resource(resource):
+    """Register a Dask resource with the UCXX context.
+
+    Register a Dask resource with the UCXX context to keep track of it, so that
+    the notifier thread and progress tasks may be stopped when no more resources
+    """
+    ctx = ucxx.core._get_ctx()
+
+    with ctx._dask_resources_lock:
+        ctx._dask_resources.add(resource)
+        ctx.start_notifier_thread()
+        ctx.continuous_ucx_progress()
+
+
+def _deregister_dask_resource(resource):
+    """Deregister a Dask resource with the UCXX context.
+
+    Deregister a Dask resource from the UCXX context, and if no resources remain
+    after deregistration, stop the notifier thread and progress tasks.
+    need UCXX.
+    """
+    if ucxx.core._ctx is None:
+        # Prevent creation of context if it was already destroyed, all
+        # registered references are already gone.
+        return
+
+    ctx = ucxx.core._get_ctx()
+
+    # Check if the attribute exists first, in tests the UCXX context may have
+    # been reset before some resources are deregistered.
+    if hasattr(ctx, "_dask_resources_lock"):
+        with ctx._dask_resources_lock:
+            try:
+                ctx._dask_resources.remove(resource)
+            except KeyError:
+                pass
+            _stop_notifier_thread_and_progress_tasks()
```

Comment:
I _think_ this function is only called here. Perhaps prefer to inline it?

It just does:
```
if len(ctx._dask_resources) == 0:
    ctx.stop_notifier_thread()
    ctx.progress_tasks.clear()
```